### ISSUE TITLE
infer correct TypedDict, based on literals

### ISSUE THREAD

**Issue by dsal3389** [astral-sh/ty#4013]
### Summary

Hi,
I am not sure if this is intentional, it feels really weird if it is

I was under the impression `ty` support infer typeddicts based on literals, and looks
like it is the case for constant values (values that ty can check at type checking time)

but this type inference doesn't work when the `ty` has no access to the real value
like function argument, or if the value is loaded from `json`

I have created a mini recreation of the issue:
https://play.ty.dev/b548e15a-ef52-44f3-bfc7-54dcd9ecc27d

the expected behavior is to have no type errors at all (in the playground example)

### Version

ty 0.0.60 (ae554ec78 2026-07-16)

**Comment by carljm** [comment]
Thanks! Yes, we should support this. Minimal code example:

```py
from typing import Literal, TypedDict

class Ok(TypedDict):
    success: Literal[True]
    result: int

class Err(TypedDict):
    success: Literal[False]

def handle(value: Ok | Err) -> None:
    if value["success"]:
        value["result"]  # error[invalid-key]: Unknown key "result" for TypedDict `Err`
```

It looks like there are two related gaps here. We support this kind of narrowing only for equality checks, not simple truthiness checks. And our equality-check support is missing support for booleans (it supports integers, strings, bytes, etc) -- this seems like just an oversight.

So we support the above case if it's `Literal[0]` and `Literal[1]` and uses `if value["success"] == 1`, but we don't support that version with bools, and we don't support any types with just a truthiness check.

Mypy is the only type checker that currently supports the truthiness check here.
